subprojects/core/src/main/java/org/gradle/api/internal/artifacts/JavaEcosystemAttributesDescriber.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import org.gradle.api.Named;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.Bundling;
+import org.gradle.api.attributes.Category;
+import org.gradle.api.attributes.DocsType;
+import org.gradle.api.attributes.LibraryElements;
+import org.gradle.api.attributes.Usage;
+import org.gradle.api.attributes.java.TargetJvmVersion;
+import org.gradle.api.internal.attributes.AttributeDescriber;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.internal.Cast;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Set;
+
+class JavaEcosystemAttributesDescriber implements AttributeDescriber {
+    private final static Set<Attribute<?>> ATTRIBUTES = ImmutableSet.of(
+        Usage.USAGE_ATTRIBUTE,
+        Category.CATEGORY_ATTRIBUTE,
+        LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE,
+        Bundling.BUNDLING_ATTRIBUTE,
+        TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE,
+        DocsType.DOCS_TYPE_ATTRIBUTE,
+        ProjectInternal.STATUS_ATTRIBUTE
+    );
+
+    @Override
+    public Set<Attribute<?>> getAttributes() {
+        return ATTRIBUTES;
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    @Override
+    public String describeAttributeSet(Map<Attribute<?>, ?> attributes) {
+        Object category = attr(attributes, Category.CATEGORY_ATTRIBUTE);
```

Comment:
:x: These variables should use the actual type instead of `Object` (`attr` already returns the right type)

It makes the code better readable and I think some other parts in the code can also be improved (see other comment).

Reply:
They can't because of desugaring: you can get _either_ a `Named` instance _or_ a `String` or any other desugared type.

Reply:
Then the type signature of ` <T> T attr(Map<Attribute<?>, ?> attributes, Attribute<T> attribute)` is wrong. It suggests that if you put in an attribute of type T, you'll get back T (while in fact you might get back String).

Reply:
Indeed, I created this method before realizing I could get another type.